README.rst
=======
toonlib
=======

NOTE
====

This library is using the unofficial api and is not recommended. There is a new version of the library that works with the official api and can be found here[https://github.com/costastf/toonapilib]


A library to interact with Eneco's toon.

Main information is cached for 30 seconds before reaching out to the api for
freshness. Assigning values to either the thermostat or the thermostat state effectively changing the temperature clears the cache so the next call will get fresh info about the
settings.

Most returned information is currently modeled as a named tuple
since they need no intelligence. The smartplugs and lights are proper objects
since they need to call the api and refresh their values. Everything else will evolve according to it's needs.

The library exposes the data that toon uses to graph its consumption both as
flow data (hourly consumption for the day) and graph data (yearly, monthly,
weekly, daily, hourly) consumption.


* Documentation: http://toonlib.readthedocs.io/en/latest/

Features
--------

* Reads values for gas, electric, temperature.
* Identifies connected hue lights and fibaro smartplugs
* Can read and set temperature and thermostat state
* Can turn lights or plugs on, off or toggle their state
* Can get consumption values from fibaro plugs
* More ...

TODO
____

* Change the caching library to a name spaced one.
* Fine tune the caching sanely across all required objects
* Properly implement caching for flow and graph data information


HISTORY.rst
.. :changelog:

History
-------

0.1 (13-04-2017)
----------------

* First release on pypi

0.2 (25-04-2017)
----------------

* Added support for turning on, off and toggling lights and smartplugs


0.3 (26-04-2017)
----------------

* Extended the info of lights and smartplugs. Added support for identification of locked state for switching them.


0.4 (26-04-2017)
----------------

* Exposed a solar object. Removed the solar values under the power attribute.


0.5 (26-04-2017)
----------------

* Added support for smokedetectors


1.0 (08-05-2017)
----------------

* Changed the way the state is retrieved.

tests/test_toonlib.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

"""
test_toonlib
----------------------------------
Tests for `toonlib` module.
"""

from betamax.fixtures import unittest


class TestToonlib(unittest.BetamaxTestCase):

    def setUp(self):
        """
        Test set up

        This is where you can setup things that you use throughout the tests. This method is called before every test.
        """
        pass

    def tearDown(self):
        """
        Test tear down

        This is where you should tear down what you've setup in setUp before. This method is called after every test.
        """
        pass


toonlib/toonlib.py
#!/usr/bin/env python
# -*- coding: UTF-8 -*-
# File: toonlib.py
"""A library overloading the api of the toon mobile app"""

import copy
import logging
import uuid

from requests import Session
from cachetools import cached, TTLCache

from .configuration import (STATES,
                            STATE_CACHING_SECONDS,
                            DEFAULT_STATE,
                            AUTHENTICATION_ERROR_STRINGS)
from .helpers import (ThermostatState,
                      Client,
                      PersonalDetails,
                      Agreement,
                      SmokeDetector,
                      PowerUsage,
                      Solar,
                      Usage,
                      ThermostatInfo,
                      Light,
                      SmartPlug,
                      Data)
from .toonlibexceptions import (InvalidCredentials,
                                UnableToGetSession,
                                IncompleteResponse,
                                InvalidThermostatState)

__author__ = '''Costas Tyfoxylos <costas.tyf@gmail.com>'''
__docformat__ = 'plaintext'
__date__ = '''13-03-2017'''

# This is the main prefix used for logging
LOGGER_BASENAME = '''toonlib'''
LOGGER = logging.getLogger(LOGGER_BASENAME)
LOGGER.addHandler(logging.NullHandler())

state_cache = TTLCache(maxsize=1, ttl=STATE_CACHING_SECONDS)


class Toon(object):
    """Model of the toon smart meter from eneco."""

    def __init__(self, username, password, state_retrieval_retry=1):
        logger_name = u'{base}.{suffix}'.format(base=LOGGER_BASENAME,
                                                suffix=self.__class__.__name__)
        self._logger = logging.getLogger(logger_name)
        self._session = Session()
        self.username = username
        self.password = password
        self.base_url = 'https://toonopafstand.eneco.nl/toonMobileBackendWeb'
        self.agreements = None
        self.agreement = None
        self.client = None
        self._state_ = DEFAULT_STATE
        self._state_retries = state_retrieval_retry
        self._uuid = None
        self.data = Data(self)
        self._login()

    def _reset(self):
        self.agreements = None
        self.agreement = None
        self.client = None
        self._state_ = DEFAULT_STATE
        self._uuid = None

    def _authenticate(self):
        """Authenticates to the api and sets up client information."""
        data = {'username': self.username,
                'password': self.password}
        url = '{base}/client/login'.format(base=self.base_url)
        response = self._session.get(url, params=data)
        data = response.json()
        if not data.get('success'):
            raise InvalidCredentials(data.get('reason', None))
        self._populate_info(data)

    def _populate_info(self, data):
        agreements = data.pop('agreements')
        self.agreements = [Agreement(agreement.get('agreementId'),
                                     agreement.get('agreementIdChecksum'),
                                     agreement.get('city'),
                                     agreement.get('displayCommonName'),
                                     agreement.get('displayHardwareVersion'),
                                     agreement.get('displaySoftwareVersion'),
                                     agreement.get('heatingType'),
                                     agreement.get('houseNumber'),
                                     agreement.get('isBoilerManagement'),
                                     agreement.get('isToonSolar'),
                                     agreement.get('isToonly'),
                                     agreement.get('postalCode'),
                                     agreement.get('street'))
                           for agreement in agreements]
        self.agreement = self.agreements[0]
        details = PersonalDetails(data.get('enecoClientNumber'),
                                  data.get('enecoClientEmailAddress'),
                                  data.get('enecoClientFirstName'),
                                  data.get('enecoClientLastName'),
                                  data.get('enecoClientMiddleName'),
                                  data.get('enecoClientMobileNumber'),
                                  data.get('enecoClientPhoneNumber'))
        self.client = Client(data.get('clientId'),
                             data.get('clientIdChecksum'),
                             data.get('passwordHash'),
                             data.get('sample'),
                             details)

    @property
    def _parameters(self):
        return {'clientId': self.client.id,
                'clientIdChecksum': self.client.checksum,
                'random': self._uuid or uuid.uuid4()}

    def _login(self):
        self._authenticate()
        self._get_session()

    def _logout(self):
        """Log out of the API."""
        url = '{base}/client/auth/logout'.format(base=self.base_url)
        response = self._session.get(url, params=self._parameters)
        if response.ok:
            self._reset()
            return True
        else:
            return False

    def _get_session(self):
        data = copy.copy(self._parameters)
        data.update({'agreementId': self.agreement.id,
                     'agreementIdChecksum': self.agreement.checksum})
        url = '{base}/client/auth/start'.format(base=self.base_url)
        response = self._session.get(url, params=data)
        if not response.ok:
            self._logout()
            message = ('\n\tStatus Code :{}'
                       '\n\tText :{}').format(response.status_code,
                                              response.text)
            raise UnableToGetSession(message)
        else:
            uuid_kpi = response.json().get('displayUuidKpi', None)
            if uuid_kpi:
                self._uuid = uuid_kpi.get('uuid', None)
        return True

    def _clear_cache(self):
        self._logger.debug('Clearing state cache.')
        state_cache.clear()

    @property
    @cached(state_cache)
    def _state(self):
        """The internal state of the object.

        The api responses are not consistent so a retry is performed on every
        call with information updating the internally saved state refreshing
        the data. The info is cached for STATE_CACHING_SECONDS.

        :return: The current state of the toons' information state.
        """
        state = {}
        required_keys = ('deviceStatusInfo',
                         'gasUsage',
                         'powerUsage',
                         'thermostatInfo',
                         'thermostatStates')
        try:
            for _ in range(self._state_retries):
                state.update(self._get_data('/client/auth/retrieveToonState'))
        except TypeError:
            self._logger.exception('Could not get answer from service.')
        message = ('Updating internal state with retrieved '
                   'state:{state}').format(state=state)
        self._logger.debug(message)
        self._state_.update(state)
        if not all([key in self._state_.keys() for key in required_keys]):
            raise IncompleteResponse(state)
        return self._state_

    def _get_data(self, endpoint, params=None):
        url = '{base}{endpoint}'.format(base=self.base_url,
                                        endpoint=endpoint)
        response = self._session.get(url, params=params or self._parameters)
        if response.status_code == 500:
            error_message = response.json().get('reason', '')
            if any([message in error_message
                    for message in AUTHENTICATION_ERROR_STRINGS]):
                self._logger.info('Detected an issue with authentication. '
                                  'Trying to reauthenticate.')
                self._login()
                return self._get_data(endpoint, params)
        elif not response.ok:
            self._logger.debug(('\n\tStatus Code :{}'
                                '\n\tText :{}').format(response.status_code,
                                                       response.text))
        else:
            try:
                return response.json()
            except (ValueError, TypeError):
                self._logger.debug(('\n\tStatus Code :{}'
                                    '\n\tText :{}').format(response.status_code,
                                                           response.text))
        return {}

    @property
    def smokedetectors(self):
        """:return: A list of smokedetector objects modeled as named tuples"""
        return [SmokeDetector(smokedetector.get('devUuid'),
                              smokedetector.get('name'),
                              smokedetector.get('lastConnectedChange'),
                              smokedetector.get('connected'),
                              smokedetector.get('batteryLevel'),
                              smokedetector.get('type'))
                for smokedetector in self._state.get('smokeDetectors',
                                                     {}).get('device', [])]

    def get_smokedetector_by_name(self, name):
        """Retrieves a smokedetector object by its name

        :param name: The name of the smokedetector to return
        :return: A smokedetector object
        """
        return next((smokedetector for smokedetector in self.smokedetectors
                     if smokedetector.name.lower() == name.lower()), None)

    @property
    def lights(self):
        """:return: A list of light objects"""
        return [Light(self, light.get('name'))
                for light in self._state.get('deviceStatusInfo',
                                             {}).get('device', [])
                if light.get('rgbColor')]

    def get_light_by_name(self, name):
        """Retrieves a light object by its name

        :param name: The name of the light to return
        :return: A light object
        """
        return next((light for light in self.lights
                     if light.name.lower() == name.lower()), None)

    @property
    def smartplugs(self):
        """:return: A list of smartplug objects."""
        return [SmartPlug(self, plug.get('name'))
                for plug in self._state.get('deviceStatusInfo',
                                            {}).get('device', [])
                if plug.get('networkHealthState')]

    def get_smartplug_by_name(self, name):
        """Retrieves a smartplug object by its name

        :param name: The name of the smartplug to return
        :return: A smartplug object
        """
        return next((plug for plug in self.smartplugs
                     if plug.name.lower() == name.lower()), None)

    @property
    def gas(self):
        """:return: A gas object modeled as a named tuple"""
        usage = self._state['gasUsage']
        return Usage(usage.get('avgDayValue'),
                     usage.get('avgValue'),
                     usage.get('dayCost'),
                     usage.get('dayUsage'),
                     usage.get('isSmart'),
                     usage.get('meterReading'),
                     usage.get('value'))

    @property
    def power(self):
        """:return: A power object modeled as a named tuple"""
        power = self._state['powerUsage']
        return PowerUsage(power.get('avgDayValue'),
                          power.get('avgValue'),
                          power.get('dayCost'),
                          power.get('dayUsage'),
                          power.get('isSmart'),
                          power.get('meterReading'),
                          power.get('value'),
                          power.get('meterReadingLow'),
                          power.get('dayLowUsage'))

    @property
    def solar(self):
        power = self._state['powerUsage']
        return Solar(power.get('maxSolar'),
                     power.get('valueProduced'),
                     power.get('valueSolar'),
                     power.get('avgProduValue'),
                     power.get('meterReadingLowProdu'),
                     power.get('meterReadingProdu'),
                     power.get('dayCostProduced'))

    @property
    def thermostat_info(self):
        """:return: A thermostatinfo object modeled as a named tuple"""
        info = self._state['thermostatInfo']
        return ThermostatInfo(info.get('activeState'),
                              info.get('boilerModuleConnected'),
                              info.get('burnerInfo'),
                              info.get('currentDisplayTemp'),
                              info.get('currentModulationLevel'),
                              info.get('currentSetpoint'),
                              info.get('currentTemp'),
                              info.get('errorFound'),
                              info.get('haveOTBoiler'),
                              info.get('nextProgram'),
                              info.get('nextSetpoint'),
                              info.get('nextState'),
                              info.get('nextTime'),
                              info.get('otCommError'),
                              info.get('programState'),
                              info.get('randomConfigId'),
                              info.get('realSetpoint'))

    @property
    def thermostat_states(self):
        """:return: A list of thermostatstate object modeled as named tuples"""
        return [ThermostatState(STATES[state.get('id')],
                                state.get('id'),
                                state.get('tempValue'),
                                state.get('dhw'))
                for state in self._state['thermostatStates']['state']]

    def get_thermostat_state_by_name(self, name):
        """Retrieves a thermostat state object by its assigned name

        :param name: The name of the thermostat state
        :return: The thermostat state object
        """
        self._validate_thermostat_state_name(name)
        return next((state for state in self.thermostat_states
                     if state.name.lower() == name.lower()), None)

    def get_thermostat_state_by_id(self, id_):
        """Retrieves a thermostat state object by its id

        :param id_: The id of the thermostat state
        :return: The thermostat state object
        """
        return next((state for state in self.thermostat_states
                     if state.id == id_), None)

    @staticmethod
    def _validate_thermostat_state_name(name):
        if name.lower() not in [value.lower() for value in STATES.values()
                                if not value.lower() == 'unknown']:
            raise InvalidThermostatState(name)

    @property
    def thermostat_state(self):
        """The state of the thermostat programming

        :return: A thermostat state object of the current setting
        """
        current_state = self.thermostat_info.active_state
        state = self.get_thermostat_state_by_id(current_state)
        if not state:
            self._logger.debug('Manually set temperature, no Thermostat '
                               'State chosen!')
        return state

    @thermostat_state.setter
    def thermostat_state(self, name):
        """Changes the thermostat state to the one passed as an argument as name

        :param name: The name of the thermostat state to change to.
        """
        self._validate_thermostat_state_name(name)
        id_ = next((key for key in STATES.keys()
                    if STATES[key].lower() == name.lower()), None)
        data = copy.copy(self._parameters)
        data.update({'state': 2,
                     'temperatureState': id_})
        response = self._get_data('/client/auth/schemeState', data)
        self._logger.debug('Response received {}'.format(response))
        self._clear_cache()

    @property
    def thermostat(self):
        """The current setting of the thermostat as temperature

        :return: A float of the current setting of the temperature of the
        thermostat
        """
        return float(self.thermostat_info.current_set_point / 100)

    @thermostat.setter
    def thermostat(self, temperature):
        """A temperature to set the thermostat to. Requires a float.

        :param temperature: A float of the desired temperature to change to.
        """
        target = int(temperature * 100)
        data = copy.copy(self._parameters)
        data.update({'value': target})
        response = self._get_data('/client/auth/setPoint', data)
        self._logger.debug('Response received {}'.format(response))
        self._clear_cache()

    @property
    def temperature(self):
        """The current actual temperature as perceived by toon.

        :return: A float of the current temperature
        """
        return float(self.thermostat_info.current_temperature / 100)


toonlib/__init__.py
# -*- coding: utf-8 -*-
"""toonlib package"""
from ._version import __version__
from .toonlibexceptions import (InvalidCredentials,
                                UnableToGetSession,
                                IncompleteResponse,
                                InvalidThermostatState)
from .configuration import (STATES,
                            STATE_CACHING_SECONDS,
                            DEFAULT_STATE,
                            AUTHENTICATION_ERROR_STRINGS)
from .helpers import (ThermostatState,
                      Client,
                      PersonalDetails,
                      Agreement,
                      SmokeDetector,
                      PowerUsage,
                      Solar,
                      Usage,
                      ThermostatInfo,
                      Light,
                      SmartPlug,
                      Data)
from .toonlib import Toon

__author__ = '''Costas Tyfoxylos'''
__email__ = '''costas.tyf@gmail.com'''

# This is to 'use' the module(s), so lint doesn't complain
assert __version__

# assert the exceptions
assert InvalidCredentials
assert UnableToGetSession
assert IncompleteResponse
assert InvalidThermostatState

# assert the objects
assert Toon
assert ThermostatState
assert Client
assert PersonalDetails
assert Agreement
assert SmokeDetector
assert PowerUsage
assert Solar
assert Usage
assert ThermostatInfo
assert Light
assert SmartPlug
assert Data

# assert configurations
assert STATES
assert STATE_CACHING_SECONDS
assert DEFAULT_STATE
assert AUTHENTICATION_ERROR_STRINGS


toonlib/helpers.py
#!/usr/bin/env python
# -*- coding: UTF-8 -*-
# File: toonlib.py
"""All helper objects will live here"""

import logging
import copy
from collections import namedtuple

__author__ = '''Costas Tyfoxylos <costas.tyf@gmail.com>'''
__docformat__ = 'plaintext'
__date__ = '''13-03-2017'''

LOGGER_BASENAME = '''helpers'''
LOGGER = logging.getLogger(LOGGER_BASENAME)
LOGGER.addHandler(logging.NullHandler())


ThermostatState = namedtuple('ThermostatState', ('name',
                                                 'id',
                                                 'temperature',
                                                 'dhw'))

ThermostatInfo = namedtuple('ThermostatInfo', ('active_state',
                                               'boiler_connected',
                                               'burner_info',
                                               'current_displayed_temperature',
                                               'current_modulation_level',
                                               'current_set_point',
                                               'current_temperature',
                                               'error_found',
                                               'have_ot_boiler',
                                               'next_program',
                                               'next_set_point',
                                               'next_state',
                                               'next_time',
                                               'ot_communication_error',
                                               'program_state',
                                               'random_configuration_id',
                                               'real_set_point'))

Usage = namedtuple('Usage', ('average_daily',
                             'average',
                             'daily_cost',
                             'daily_usage',
                             'is_smart',
                             'meter_reading',
                             'value'))

Low = namedtuple('Low', ('meter_reading_low', 'daily_usage_low'))

Solar = namedtuple('Solar', ('maximum',
                             'produced',
                             'value',
                             'average_produced',
                             'meter_reading_low_produced',
                             'meter_reading_produced',
                             'daily_cost_produced'))

PowerUsage = namedtuple('PowerUsage',
                        Usage._fields + Low._fields)

Agreement = namedtuple('Agreement', ('id',
                                     'checksum',
                                     'city',
                                     'display_common_name',
                                     'display_hardware_version',
                                     'display_software_version',
                                     'heating_type',
                                     'house_number',
                                     'boiler_management',
                                     'solar',
                                     'toonly',
                                     'post_code',
                                     'street_name'))

PersonalDetails = namedtuple('PersonalDetails', ('number',
                                                 'email',
                                                 'first_name',
                                                 'last_name',
                                                 'middle_name',
                                                 'mobile_number',
                                                 'phone_number'))

Client = namedtuple('Client', ('id',
                               'checksum',
                               'hash',
                               'sample',
                               'personal_details'))

SmokeDetector = namedtuple('SmokeDetector', ('device_uuid',
                                             'name',
                                             'last_connected_change',
                                             'is_connected',
                                             'battery_level',
                                             'device_type'))


class Data(object):
    """Data object exposing flow and graph attributes."""

    class Flow(object):
        """The object that exposes the flow information of categories in toon

        The information is rrd metrics and the object dynamically handles the
        accessing of attributes matching with the corresponding api endpoint
        if they are know, raises an exception if not.
        """

        def __init__(self, toon_instance):
            self.toon = toon_instance
            self._endpoint = {'power': '/client/auth/getElecFlowData',
                              'gas': '/client/auth/getGasFlowData',
                              'solar': '/client/auth/getSolarFlowData'}

        def __getattr__(self, name):
            """Implements dynamic atributes on the Flow object"""
            endpoint = self._endpoint.get(name)
            if not endpoint:
                raise AttributeError(name)
            data = self.toon._get_data(endpoint)
            return data.get('graphData') if data.get('success') else None

    class Graph(object):
        """The object that exposes the graph information of categories in toon

        The information is rrd metrics and the object dynamically handles the
        accessing of attributes matching with the corresponding api endpoint
        if they are know, raises an exception if not.
        """

        def __init__(self, toon_instance):
            self.toon = toon_instance
            self._endpoint = {'power': '/client/auth/getElecGraphData',
                              'gas': '/client/auth/getGasGraphData',
                              'solar': '/client/auth/getSolarGraphData',
                              'district_heat':
                                  '/client/auth/getDistrictHeatGraphData'}

        def __getattr__(self, name):
            """Implements dynamic atributes on the Graph object"""
            endpoint = self._endpoint.get(name)
            if not endpoint:
                raise AttributeError(name)
            data = self.toon._get_data(endpoint)
            return data.get('graphData') if data.get('success') else None

    def __init__(self, toon_instance):
        logger_name = u'{base}.{suffix}'.format(base=LOGGER_BASENAME,
                                                suffix=self.__class__.__name__)
        self._logger = logging.getLogger(logger_name)
        self.flow = self.Flow(toon_instance)
        self.graph = self.Graph(toon_instance)


class Switch(object):
    """Core object to implement the turning on, off or toggle

    Both hue lamps and fibaro plugs have a switch component that is shared.
    This implements that usage.
    """

    def __init__(self, toon_instance, name):
        logger_name = u'{base}.{suffix}'.format(base=LOGGER_BASENAME,
                                                suffix=self.__class__.__name__)
        self._logger = logging.getLogger(logger_name)
        self.toon = toon_instance
        self._name = name

    @property
    def name(self):
        return self._name

    def _get_value(self, name, config=False):
        key = 'deviceConfigInfo' if config else 'deviceStatusInfo'
        return next((item.get(name) for item in
                     self.toon._state.get(key).get('device')  # noqa
                     if item.get('name') == self.name), None)

    def toggle(self):
        return self._change_state(not self.current_state)

    def turn_on(self):
        return self._change_state(1)

    @property
    def status(self):
        return 'on' if self.current_state else 'off'

    def _change_state(self, state):
        if not self.can_toggle:
            self._logger.warning('The item is not connected or locked, cannot '
                                 'change state.')
            return False
        else:
            data = copy.copy(self.toon._parameters)  # noqa
            data.update({'state': int(state),
                         'devUuid': self.device_uuid})
            response = self.toon._get_data('/client/auth/smartplug/setTarget',  # noqa
                                           data)
            self._logger.debug('Response received {}'.format(response))
            self.toon._clear_cache()  # noqa
            return True

    @property
    def can_toggle(self):
        if not self.is_connected or self.is_locked:
            return False
        else:
            return True

    def turn_off(self):
        return self._change_state(0)

    @property
    def device_uuid(self):
        return self._get_value('devUUID')

    @property
    def is_connected(self):
        value = self._get_value('isConnected')
        return True if value else False

    @property
    def current_state(self):
        return self._get_value('currentState')

    @property
    def device_type(self):
        return self._get_value('devType', config=True)

    @property
    def in_switch_all_group(self):
        value = self._get_value('inSwitchAll', config=True)
        return True if value else False

    @property
    def in_switch_schedule(self):
        value = self._get_value('inSwitchSchedule', config=True)
        return True if value else False

    @property
    def zwave_index(self):
        return self._get_value('position', config=True)

    @property
    def is_locked(self):
        value = self._get_value('switchLocked', config=True)
        return True if value else False

    @property
    def zwave_uuid(self):
        return self._get_value('zwUuid', config=True)


class SmartPlug(Switch):
    """Object modeling the fibaro smart plugs the toon can interact with.

    It inherits from switch which is the common interface with the hue
    lamps to turn on, off or toggle
    """

    def __init__(self, toon_instance, name):
        super(SmartPlug, self).__init__(toon_instance, name)

    @property
    def average_usage(self):
        return self._get_value('avgUsage')

    @property
    def current_usage(self):
        return self._get_value('currentUsage')

    @property
    def daily_usage(self):
        return self._get_value('dayUsage')

    @property
    def network_health_state(self):
        return self._get_value('networkHealthState')

    @property
    def usage_capable(self):
        value = self._get_value('usageCapable', config=True)
        return True if value else False

    @property
    def quantity_graph_uuid(self):
        return self._get_value('quantityGraphUuid', config=True)

    @property
    def flow_graph_uuid(self):
        return self._get_value('flowGraphUuid', config=True)


class Light(Switch):
    """Object modeling the hue light bulbs that toon can interact with.

    It inherits from switch which is the common interface with the hue
    lamps to turn on, off or toggle
    """

    def __init__(self, toon_instance, name):
        super(Light, self).__init__(toon_instance, name)

    @property
    def rgb_color(self):
        return self._get_value('rgbColor')
